src/looprestoration.rs

```diff
@@ -1549,148 +1549,103 @@ mod neon {
+            bd: BD,
+        ) {
+            let left = left
+                .map(|left| left.as_ptr().cast())
+                .unwrap_or_else(ptr::null);
+            let src = src.cast();
+            let bd = bd.into_c();
+            // SAFETY: asm should be safe.
+            unsafe { self.get()(sumsq, sum, left, src, w, edges, bd) }
         }
-
-        bd_fn!(asm_fn, BD, sgr_box5_row_h, neon)(
-            sumsq,
-            sum,
-            left.map(|left| left.as_ptr().cast())
-                .unwrap_or_else(ptr::null),
-            src.cast(),
-            w,
-            edges,
-            bd.into_c(),
-        )
     }
 
-    unsafe fn rav1d_sgr_box35_row_h_neon<BD: BitDepth>(
+    wrap_fn_ptr!(unsafe extern "C" fn sgr_box35_row_h(
         sumsq3: *mut i32,
         sum3: *mut i16,
         sumsq5: *mut i32,
         sum5: *mut i16,
-        left: Option<&[LeftPixelRow<BD::Pixel>]>,
-        src: *const BD::Pixel,
+        left: *const LeftPixelRow<DynPixel>,
+        src: *const DynPixel,
         w: c_int,
         edges: LrEdgeFlags,
-        bd: BD,
-    ) {
-        macro_rules! asm_fn {
-            (fn $name:ident) => {{
-                extern "C" {
-                    fn $name(
-                        sumsq3: *mut i32,
-                        sum3: *mut i16,
-                        sumsq5: *mut i32,
-                        sum5: *mut i16,
-                        left: *const LeftPixelRow<DynPixel>,
-                        src: *const DynPixel,
-                        w: c_int,
-                        edges: LrEdgeFlags,
-                        bitdepth_max: c_int,
-                    );
-                }
-                $name
-            }};
-        }
-        bd_fn!(asm_fn, BD, sgr_box35_row_h, neon)(
-            sumsq3,
-            sum3,
-            sumsq5,
-            sum5,
-            left.map(|left| left.as_ptr().cast())
-                .unwrap_or_else(ptr::null),
-            src.cast(),
-            w,
-            edges,
-            bd.into_c(),
-        )
-    }
+        bitdepth_max: c_int,
+    ) -> ());
 
-    extern "C" {
-        fn dav1d_sgr_box3_vert_neon(
-            sumsq: *mut *mut i32,
-            sum: *mut *mut i16,
-            AA: *mut i32,
-            BB: *mut i16,
+    impl sgr_box35_row_h::Fn {
+        fn call<BD: BitDepth>(
+            &self,
+            sumsq3: *mut i32,
+            sum3: *mut i16,
+            sumsq5: *mut i32,
+            sum5: *mut i16,
+            left: Option<&[LeftPixelRow<BD::Pixel>]>,
+            src: *const BD::Pixel,
             w: c_int,
-            s: c_int,
-            bitdepth_max: c_int,
-        );
+            edges: LrEdgeFlags,
+            bd: BD,
+        ) {
+            let left = left
+                .map(|left| left.as_ptr().cast())
+                .unwrap_or_else(ptr::null);
+            let src = src.cast();
+            let bd = bd.into_c();
+            // SAFETY: asm should be safe.
```

Comment:
Not sure what this means. _Should_ be safe? According to what? We seem to have this safety comment in a lot of places. I think this should just be something like "asm call."

Reply:
Yeah this is the safety comment I've been using for unconditionally asm (no fallback) calls.  We depend on the asm being safe for soundness which is why I used "should", but I'm open to different wording, though I'd want to change them all en masse if I do.

Reply:
Alright, let's discuss, "should" isn't clear here. Agreed we should do that in a separate change.